app/src/main/java/org/dharmaseed/android/DBManager.java

```diff
@@ -134,6 +138,42 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
             db.execSQL(C.TalkHistory.CREATE_TABLE);
             Log.i(LOG_TAG,"Upgrade: Created talk history table");
         }
+
+        File dbUpgradeFile = context.getDatabasePath(DB_NAME+".upgrade");
+        try {
+            copyAssetDB(dbUpgradeFile);
+            String dbUpgradePath = dbUpgradeFile.getAbsolutePath();
+            SQLiteDatabase assetDb = SQLiteDatabase.openDatabase(dbUpgradePath, null, SQLiteDatabase.OPEN_READONLY);
+            int assetVersion = assetDb.getVersion();
+            assetDb.close();
+            Log.d(LOG_TAG, "Asset DB has version " + assetVersion);
+            if (assetVersion == newVersion) {
+                Log.i(LOG_TAG, "Asset DB version matches upgrade version. Copying updated tables!");
+                db.execSQL("ATTACH DATABASE '" + dbUpgradePath + "' AS upgradeDb");
+                String[] tables = {
+                        C.Talk.TABLE_NAME,
+                        C.Teacher.TABLE_NAME,
+                        C.Center.TABLE_NAME,
+                        C.Edition.TABLE_NAME
+                };
+                for (String table: tables) {
+                    db.execSQL("DROP TABLE IF EXISTS " + table);
+                    db.execSQL("CREATE TABLE " + table + " AS SELECT * FROM upgradeDb." + table);
+                    Log.i(LOG_TAG, "Copied table " + table + " from asset DB.");
+                }
+                db.setTransactionSuccessful();
+                db.endTransaction();
+                db.execSQL("DETACH upgradeDb");
+                db.beginTransaction();
```

Comment:
Ah yes, that's a good question. My code originally did not contain any of the transaction stuff, either, and it would have stayed like that if only it had worked!

The reason why this is necessary is that `onUpgrade` is apparently being run in the following way:

`
db.beginTransaction();
   try {
     onUpgrade(...);
     db.setTransactionSuccessful();
   } finally {
     db.endTransaction();
   }
`

This is to ensure that the upgrade can be rolled back in case something goes badly wrong, which makes sense. It also means that there must be an open transaction when `onUpgrade` returns.

The clean solution would have been to just leave the transaction open, of course. But that does not work  because the `DETACH` command must run outside of a transaction. So the solution is to close the original transaction and re-open a new one after the `DETACH`. This is not 100% clean, but I think we should be OK as long as this DB update code runs at the very end of `onUpgrade`, such that we are sure that nothing will go wrong after we successfully end the first transaction.

Reply:
Ah that makes a lot of sense, thanks! Would you mind adding a brief comment here outlining that explanation? I think it might be helpful for understanding the code in the future, since it's not immediately obvious that `onUpgrade` already is running within a transaction.

Reply:
The new comments look great, thank you!